pkg/storage/redis.go

```diff
@@ -62,9 +77,51 @@ func (b *RedisDB) Close() error {
 	return b.db.Close()
 }
 
+func (b *RedisDB) Execute(ctx context.Context, businessLogicFunc BusinessLogicFunc, watchKeys []WatchKey) (any, error) {
+	var finalOutput any
+	// Transactional function.
+	txf := func(tx *goredislib.Tx) error {
+		// Operation is commited only if the watched keys remain unchanged.
+		_, err := tx.TxPipelined(ctx, func(pipe goredislib.Pipeliner) error {
+			redisTx := redisTx{pipe}
+			var err error
+
+			finalOutput, err = businessLogicFunc(ctx, &redisTx)
+			if err != nil {
+				return err
+			}
+			return nil
+		})
+		return err
+	}
+
+	watchKeysStr := make([]string, 0)
+
+	for _, wc := range watchKeys {
+		watchKeysStr = append(watchKeysStr, fmt.Sprintf("%v:%v", wc.Namespace, wc.Key))
+	}
+
+	expBackoff := backoff.NewExponentialBackOff()
+	expBackoff.MaxElapsedTime = MaxElapsedTime
+
+	err := backoff.Retry(func() error {
+		err := b.db.Watch(ctx, txf, watchKeysStr...)
+		if err != nil && errors.Is(err, goredislib.TxFailedErr) {
+			logrus.Warn("Optimistic lock lost. Retrying..")
+		}
+		return err
```

Comment:
Perhaps we shouldn't retry when the error is different than `TxFailedErr`. You can do so as follows:

```suggestion
		if err != nil && errors.Is(err, goredislib.TxFailedErr) {
			logrus.Warn("Optimistic lock lost. Retrying..")
			return err
		}
		return backoff.Permanent(err)
```

Reply:
ok that makes sense,